core/core_test.go

```diff
@@ -75,6 +82,190 @@ func TestTranscode(t *testing.T) {
+
+	transcodeLoopTimeout = 100 * time.Millisecond
+	require := require.New(t)
+
+	sessionIDs := []string{"first session ID", "second session ID"}
+	n.pmSessionsMutex.Lock()
+	n.PMSessions[md.ManifestID] = make(map[string]bool)
+	n.PMSessions[md.ManifestID][sessionIDs[0]] = true
+	n.PMSessions[md.ManifestID][sessionIDs[1]] = true
+	n.pmSessionsMutex.Unlock()
+
+	// Need to use this because the order of the string slice used to call this function
+	// later cannot be guaranteed, since they are coming from a map's keys.
+	argsMatcher := mock.MatchedBy(func(IDs []string) bool {
+		sort.Strings(sessionIDs)
+		sort.Strings(IDs)
+
+		return reflect.DeepEqual(sessionIDs, IDs)
+	})
+
+	recipient.On("RedeemWinningTickets", argsMatcher).Return(nil)
+
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+
+	recipient.AssertExpectations(t)
+}
+
+func TestTranscodeLoop_GivenNoPMSession_DoesntTryToRedeem(t *testing.T) {
+	recipient := new(pm.MockRecipient)
+	//Set up the node
+	drivers.NodeStorage = drivers.NewMemoryDriver("")
+	db, _ := common.InitDB("file:TestTranscode?mode=memory&cache=shared")
+	defer db.Close()
+	seth := &eth.StubClient{}
+	tmp, _ := ioutil.TempDir("", "")
+	n, _ := NewLivepeerNode(seth, tmp, db)
+	n.Recipient = recipient
+	defer os.RemoveAll(tmp)
+	ffmpeg.InitFFmpeg()
+	ss := StubSegment()
+	md := &SegTranscodingMetadata{Profiles: videoProfiles}
+	n.Transcoder = NewLocalTranscoder(tmp)
+	transcodeLoopTimeout = 100 * time.Millisecond
+	require := require.New(t)
+	recipient.On("RedeemWinningTickets", mock.Anything).Return(nil)
+
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+
+	recipient.AssertNotCalled(t, "RedeemWinningTickets", mock.Anything)
+}
+
+func TestTranscodeLoop_GivenRedeemError_ErrorLogIsWritten(t *testing.T) {
+	recipient := new(pm.MockRecipient)
+	//Set up the node
+	drivers.NodeStorage = drivers.NewMemoryDriver("")
+	db, _ := common.InitDB("file:TestTranscode?mode=memory&cache=shared")
+	defer db.Close()
+	seth := &eth.StubClient{}
+	tmp, _ := ioutil.TempDir("", "")
+	n, _ := NewLivepeerNode(seth, tmp, db)
+	n.Recipient = recipient
+	defer os.RemoveAll(tmp)
+	ffmpeg.InitFFmpeg()
+	ss := StubSegment()
+	md := &SegTranscodingMetadata{Profiles: videoProfiles}
+	n.Transcoder = NewLocalTranscoder(tmp)
+	transcodeLoopTimeout = 100 * time.Millisecond
+	require := require.New(t)
+
+	sessionID := "some session ID"
+	n.pmSessionsMutex.Lock()
+	n.PMSessions[md.ManifestID] = make(map[string]bool)
+	n.PMSessions[md.ManifestID][sessionID] = true
+	n.pmSessionsMutex.Unlock()
+	recipient.On("RedeemWinningTickets", mock.Anything).Return(fmt.Errorf("some error"))
+
+	errorLogsBefore := glog.Stats.Error.Lines()
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+	errorLogsAfter := glog.Stats.Error.Lines()
+
+	assert.Equal(t, int64(1), errorLogsAfter-errorLogsBefore)
```

Comment:
This test seems fragile - if someone ever adds an additional log to `sendToTranscodeLoop` this test would fail. Another question: is it worthwhile to test the presence of log messages in this manner?

Reply:
Excellent question. I was thinking what's the best way to handle a failure in redeeming winning tickets, and currently logging an error was the best I could come up with (perhaps we want to throw some event to some monitoring solution, but I think that's for a separate discussion about instrumentation).

If we agree that logging upon redemption error is the best course of action, I felt like erring on the side of having a test for it rather than not, because without a test, for all I know we are doing nothing in case of an error.

I agree this test is somewhat fragile, but thought we might want to keep it as long as it's not causing us too much trouble, for the benefit of assurance that we're at least doing something about failed redemptions.

Wdyt?

Reply:
That makes sense and I think emitting some type of event for a monitoring solution would be a nice future addition with logging being the short term solution.

Reply:
Yes! Would love for us to take monitoring to the next level!